fetcher/fedora.go

```diff
@@ -0,0 +1,128 @@
+	for _, v := range target {
+		reqs = append(reqs, fetchRequest{
+			target:       v,
+			url:          fmt.Sprintf(t, v),
+			bzip2:        false,
+			xz:           false,
+			concurrently: false,
+		})
+	}
+	return
+}
+
+// FetchFedora fetch OVAL from Fedora
+func FetchUpdateInfosFedora(versions []string) (map[string]FedoraUpdates, error) {
+	feeds, err := fetchFeedFilesFedora(versions)
+	if err != nil {
+		return nil, err
+	}
+
+	updates, err := fetchUpdateInfosFedora(feeds)
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := parseFetchResultsFedora(updates)
+	if err != nil {
+		return nil, err
+	}
+	return results, nil
+}
+
+func fetchFeedFilesFedora(versions []string) ([]FetchResult, error) {
+	reqs := newFedoraFetchRequests(versions)
+	if len(reqs) == 0 {
+		return nil, fmt.Errorf("There are no versions to fetch")
+	}
+	results, err := fetchFeedFiles(reqs)
+	if err != nil {
+		return nil, xerrors.Errorf("Failed to fetch. err: %w", err)
+	}
+	return results, nil
+}
+
+func fetchUpdateInfosFedora(results []FetchResult) ([]FetchResult, error) {
+	// extract updateinfo
+	var updateInfoReqs []fetchRequest
+	for _, r := range results {
+		var repoMd RepoMd
+		if err := xml.NewDecoder(bytes.NewBuffer(r.Body)).Decode(&repoMd); err != nil {
+			log15.Warn(fmt.Sprintf("Failed to decode repomd. Skip to fetch version %s", r.Target), "err", err)
+			continue
+		}
+
+		for _, repo := range repoMd.RepoList {
+			if repo.Type == "updateinfo" {
+				u, err := url.Parse(r.URL)
+				if err != nil {
+					return nil, err
+				}
+				u.Path = strings.Replace(u.Path, "repodata/repomd.xml", repo.Location.Href, 1)
+				req := fetchRequest{
+					url:          u.String(),
+					target:       r.Target,
+					bzip2:        false,
+					xz:           true,
+					concurrently: false,
+				}
+				updateInfoReqs = append(updateInfoReqs, req)
+				break
+			}
+		}
+	}
+
+	if len(updateInfoReqs) == 0 {
+		return nil, fmt.Errorf("No updateinfo field in the repomd")
+	}
+
+	results, err := fetchFeedFiles(updateInfoReqs)
+	if err != nil {
+		return nil, xerrors.Errorf("Failed to fetch. err: %w", err)
+	}
+	return results, nil
+}
+
+func parseFetchResultsFedora(results []FetchResult) (map[string]FedoraUpdates, error) {
+	updateInfos := make(map[string]FedoraUpdates, len(results))
+	for _, r := range results {
+		var updateInfo FedoraUpdates
+		if err := xml.NewDecoder(bytes.NewReader(r.Body)).Decode(&updateInfo); err != nil {
+			return nil, err
+		}
+		var securityUpdate []FedoraUpdateInfo
+		for _, update := range updateInfo.UpdateList {
+			if update.Type == "security" {
+				cveIDs := []string{}
+				for _, ref := range update.References {
+					id := util.CveIDPattern.FindString(ref.Title)
+					if id != "" {
+						cveIDs = append(cveIDs, id)
+					}
+				}
+				if len(cveIDs) != 0 {
+					update.CVEIDs = cveIDs
+					securityUpdate = append(securityUpdate, update)
+				}
+			}
```

Comment:
In some cases, CVE-ID appears multiple times in ref.Title.
```xml
<update from="updates@fedoraproject.org" status="stable" type="security" version="2.0">
    <id>FEDORA-2021-5ffabdc080</id>
    <title>chromium-94.0.4606.81-1.fc35</title>
    <references>
        <reference href="https://bugzilla.redhat.com/show_bug.cgi?id=2012664" id="2012664" type="bugzilla" title="CVE-2021-37977 CVE-2021-37978 CVE-2021-37979 CVE-2021-37980 chromium: various flaws [fedora-all]"/>
    </references>
<update>

<update from="updates@fedoraproject.org" status="stable" type="security" version="2.0">
    <id>FEDORA-2021-46dc82116b</id>
    <title>community-mysql-8.0.27-1.fc35</title>
    <references>
        <reference href="https://bugzilla.redhat.com/show_bug.cgi?id=2016141" id="2016141" type="bugzilla" title="CVE-2021-2478 CVE-2021-2479 CVE-2021-2481 CVE-2021-35546 CVE-2021-35575 CVE-2021-35577 CVE-2021-35591 CVE-2021-35596 CVE-2021-35597 CVE-2021-35602 CVE-2021-35604 CVE-2021-35607 CVE-2021-35608 CVE-2021-35610 ... community-mysql: various flaws [fedora-all]"/>
    </references>
<update
```

Reply:
Sometimes there are duplicate CVE-IDs, and it is necessary to retrieve the unique CVE-ID. 
```xml
<update from="updates@fedoraproject.org" status="stable" type="security" version="2.0">
    <id>FEDORA-2021-0d4b58060d</id>
    <title>lapack-3.10.0-4.fc35</title>
    <references>
      <reference href="https://bugzilla.redhat.com/show_bug.cgi?id=2024358" id="2024358" type="bugzilla" title="CVE-2021-4048 lapack: Out-of-bounds read in *larrv"/>
      <reference href="https://bugzilla.redhat.com/show_bug.cgi?id=2029851" id="2029851" type="bugzilla" title="CVE-2021-4048 lapack: Out-of-bounds read in *larrv [fedora-all]"/>
    </references>
<update>
```

Reply:
thanks
I've fixed to retrieve all CVE-IDs from title and distinct array of CVE-IDs.

Reply:
thanks
I've fixed to retrieve all CVE-IDs from title and distinct array of CVE-IDs.